Support OpenVX delegate for several NPU/GPU

ayakoakasaka opened the pull request:
Add VX delegation as an external delegation of TFLite. By using it, several NPU/GPU (from VeriSilicon, NXP, Amlogic) can be controlled via WASI-NN.  Test Code can work with X86 simulator

sax1johno committed:
Update request.ts wasm_response_send signature (#2122)

Update wasm_response_send function signature to match api by returning
bool instead of void (was causing assemblyscript example to break)

hasheddan committed:
Update Zephyr docs to remove unsupported west subcommand (#2128)

Removes the `west espressif install` command which is no longer
supported.

https://github.com/zephyrproject-rtos/hal_espressif/pull/130

Also adds a reference to the Zephyr SDK, which will install the
necessary toolchains for all supported targets.

Signed-off-by: Daniel Mangum <georgedanielmangum@gmail.com>

yamt committed:
Update messages/comments to refer the new place of the version definition (#2133)

yamt committed:
build_wamr_lldb.yml: sync lldb build options between ubuntu and macos (#2132)

Notably, this disables python scripting support on ubuntu:

* It isn't necessary for the primary purpose of this build. (that is, the vscode extension)
* It has been disabled in the macOS counterpart.
* It allows to use the same binary for 22.04 and 20.04. (note: 22.04 and 20.04 provide different versions of libpython)

yamt committed:
build_wamr_vscode_ext.yml: vsce publish only on the official repo (#2130)

This makes it simpler to test workflow files on a fork.

wenyongh committed:
Implement Fast JIT multi-threading feature (#2134)

- Translate all the opcodes of threads spec proposal for Fast JIT
- Add the atomic flag for Fast JIT load/store IRs to support atomic load/store
- Add new atomic related Fast JIT IRs and translate them in the codegen
- Add suspend_flags check in branch opcodes and before/after call function
- Modify CI to enable Fast JIT multi-threading test

Co-authored-by: TianlongLiang <tianlong.liang@intel.com>

yamt committed:
VSCode-Extension: Download lldb built for ubuntu 20.04 (#2139)

This should allow users to use the vscode extension on ubuntu 20.04.

After https://github.com/bytecodealliance/wasm-micro-runtime/pull/2132 ,
our lldb binary for 20.04 works on ubuntu 22.04 as well.

On the other hand, lldb for 22.04 has no chance to work on ubuntu 20.04.
(because of glibc version requirement.)

ayakoakasaka committed:
Support OpenVX delegate for several NPU/GPU

Zzzabiyaka committed:
CI: Add ubsan checks to samples/wasm-c-api (#2147)

This PR adds ubsan checks with no alignment enabled to basic CI tests for
samples/wasm-c-api.

Co-authored-by: Maksim Litskevich <makslit@amazon.co.uk>

tonibofarull commented on core/iwasm/libraries/wasi-nn/README.md:
````diff
@@ -71,6 +72,18 @@ docker run \
+* vx-delegate for NPU (x86 simulater)
+
+```
+docker run \
+    -v $PWD/core/iwasm/libraries/wasi-nn/test:/assets wasi-nn-vx-delegate \
+    --dir=/assets \
+    --env="TARGET=gpu" \
+    /assets/test_tensorflow.wasm
````
is the test running in vx-delegate with `--env="TARGET=gpu"`?

ayakoakasaka replied:
So far, yes. This is the question where the external delegation should be categorised

tonibofarull commented on core/iwasm/libraries/wasi-nn/src/wasi_nn_tensorflowlite.cpp:
```diff
@@ -206,6 +210,17 @@ tensorflowlite_init_execution_context(void *tflite_ctx, graph g,
+#elif defined(WASI_NN_ENABLE_EXTERNAL_DELEGATE)
+            NN_WARN_PRINTF("external delegation enabled.");
+            auto options = TfLiteExternalDelegateOptionsDefault(
+                "/usr/lib/libvx_delegate.so");
```
Can we make this parameterizable? Other devices will also use the external delegate functionality.

ayakoakasaka replied:
Add compile option as below

wenyongh committed:
Fix ems allocator unaligned memory access on riscv64 (#2140)

Make `hmu_tree_node` struct packed and add 4 padding bytes before `kfc_tree_root_buf`
field in `gc_heap_struct` struct to ensure the `left/right/parent` fields in `hmu_tree_node`
are 8-byte aligned on the 64-bit target which doesn't support unaligned memory access.

Fix the issue reported in #2136.

g0djan committed:
libc_wasi_wrapper.c: Fix min func issue for size_t < 8 bytes on some platforms (#2152)

According to the 1999 ISO C standard (C99), size_t is an unsigned integer type of
at least 16 bit (see sections 7.17 and 7.18.3), it may be uint32 in 32-bit platforms:
https://en.cppreference.com/w/cpp/types/size_t

Calling function `size_t min(size_t, size_t)` with two uint64 arguments may get
invalid result.

Co-authored-by: Georgii Rylov <godjan@amazon.co.uk>

ayakoakasaka committed:
Set external delegate path from config

ayakoakasaka commented on core/iwasm/libraries/wasi-nn/test/Dockerfile.vx-delegate:
```diff
@@ -86,8 +86,10 @@ WORKDIR /home/wamr/core/iwasm/libraries/wasi-nn/test/build
+    -DCMAKE_INCLUDE_PATH=/usr/local/include/ \
     -DWAMR_BUILD_WASI_NN=1 \
     -DWAMR_BUILD_WASI_NN_ENABLE_EXT=1 \
+    -DWASI_NN_EXT_DELEGATE_PATH="/usr/lib/libvx_delegate.so" \
```
I added the external delegate library path from the Config

ayakoakasaka committed:
Avoid re-installing if Tensorflow is already installed for WASI-NN (#2148)

Since the Tensorflow library is already installed in many cases(especially in the
case of the embedded system), move the installation code to find_package.

TianlongLiang committed:
CI: More precise trigger paths for github actions (#2157)

Only trigger the necessary github actions when a CI file is changed.

yamt committed:
wamrc: Add --stack-usage option (#2158)

ayakoakasaka committed:
Avoid overwrite CMAKE options

ayakoakasaka committed:
Support OpenVX delegate for several NPU/GPU

ayakoakasaka committed:
Set external delegate path from config

ayakoakasaka committed:
Merge branch 'wasi-nn/vx_delegate' of https://github.com/midokura/wasm-micro-runtime into wasi-nn/vx_delegate

ayakoakasaka commented:
Thank you @tonibofarull, I create same PR to upstream
